### PR TITLE
 feat: get products details using get smarter api client

### PR DESCRIPTION
**Pull request by AfaqShuaib09**
[PROD-3116](https://2u-internal.atlassian.net/browse/PROD-3116)

### Overview:
This PR covers the get products details from getsmarter API using the getsmarter API client. It also covers the generation oauth2 token for the getsmarter API client. The access token is generated by the getsmarter API and is valid for 24 hours plus it has a functionality to get cache the access token.

#### Todos
- [x] Add environment variables to the `edx-internal` repo for getsmarter API (both stage and prod)
   https://github.com/edx/edx-internal/pull/7713 

#### Testing Instructions
- To test it locally set the following variables in your `private.py` file
    - `GETSMARTER_CLIENT_ID`
    - `GETSMARTER_CLIENT_SECRET`
    - `GETSMARTER_API_URL`
    - `GETSMARTER_PRODUCTS_DETAILS_URL`
    - `GETSMARTER_PROVIDER_URL`
    
 ### Skipped tests
- `CourseViewSetTests::test_list_query` (flaky test need to investigate its failure reason)
> The potential reason might be elasticsearch failure and return 0 records with status 400

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @AfaqShuaib09!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by mphilbrick211** [comment]
Hi @AfaqShuaib09! Just checking to see if this is ready for review? If so, you'll need to resolve the branch date error first. Thanks!

File: course_discovery/apps/course_metadata/tests/test_utils.py [hunk 1]
```diff
@@ -896,3 +899,127 @@ def test_download_and_save_course_image__for_organization_logo_override_using_dr
...
+        """
+        responses.add(
+            responses.GET,
+            settings.PRODUCT_API_URL + '/?detail=2',
+            body=json.dumps(self.SUCCESS_API_RESPONSE),
+            status=200,
+        )
+        return self.SUCCESS_API_RESPONSE
+
+    @mock.patch('course_discovery.apps.course_metadata.utils.GetSmarterEnterpriseApiClient')
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tests/test_utils.py:1007]: responses.activate is missing.

File: course_discovery/apps/course_metadata/utils.py [hunk 2]
```diff
@@ -880,3 +881,33 @@ def transform_skills_data(skills_data):
         }
         skills.append(skill_dict)
     return skills
+
+
+def get_products_details_using_getsmarter_client(getsmarter_client_credentials):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py]: nit: fetch_getsmarter_products.

File: course_discovery/apps/course_metadata/utils.py [hunk 3]
```diff
@@ -880,3 +881,33 @@ def transform_skills_data(skills_data):
...
+    getsmarter_api_client = GetSmarterEnterpriseApiClient(
+        client_id=getsmarter_client_credentials['CLIENT_ID'],
+        client_secret=getsmarter_client_credentials['CLIENT_SECRET'],
+        provider_url=getsmarter_client_credentials['PROVIDER_URL'], api_url=getsmarter_client_credentials['API_URL']
+    )
+    try:
+        response = getsmarter_api_client.request(method=method,
+                                                 url=getsmarter_client_credentials['PRODUCTS_DETAILS_URL'])
+        response.raise_for_status()
+
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py]: unnecessary empty line.

File: course_discovery/apps/course_metadata/utils.py [hunk 4]
```diff
@@ -880,3 +881,33 @@ def transform_skills_data(skills_data):
...
+        response = response.json()
+
+        if 'products' in response:
+            logger.info(f"Products found in API response: {response}")
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py]: response dict will be very large. Ideally, print the length of response.

**Review by DawoudSheraz — COMMENTED** [review]
Left a few comments.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @AfaqShuaib09!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py [hunk 5]
```diff
@@ -257,6 +257,7 @@ def test_list(self):
             self.serialize_course(Course.objects.all(), many=True)
         )
 
+    @pytest.mark.skip(reason='elasticsearch has some sort of failure and returns 0 records with status 400')
```
Comment (ansabgillani) [inline course_discovery/apps/api/v1/tests/test_views/test_courses.py]: use `unittest.skip`

**Review by ansabgillani — APPROVED** [review]

File: course_discovery/apps/course_metadata/utils.py [hunk 6]
```diff
@@ -882,3 +883,32 @@ def transform_skills_data(skills_data):
         }
         skills.append(skill_dict)
     return skills
+
+
+def fetch_getsmarter_products(getsmarter_client_credentials):
+    """ Returns the products details for the getsmarter API """
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/utils.py]: nit: from the getsmarter API

File: course_discovery/apps/course_metadata/utils.py [hunk 7]
```diff
@@ -882,3 +883,32 @@ def transform_skills_data(skills_data):
...
+            logger.info(f"Products found in API response: {len(response)}")
+            products = response['products']
+            return products
+        else:
+            logger.info(f"No products found in API response: {len(response)}")
+
+    except requests.exceptions.HTTPError as ex:
+        logger.error(f"Failed to retrieve products from getsmarter API: {ex}")
+    except Exception as ex:  # pylint: disable=broad-except
+        logger.error(f"Failed to retrieve products from getsmarter API: {ex}")
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/utils.py]: suggestion: can these messages be a bit different?

File: course_discovery/apps/course_metadata/utils.py [hunk 8]
```diff
@@ -882,3 +883,32 @@ def transform_skills_data(skills_data):
...
+
+
+def fetch_getsmarter_products(getsmarter_client_credentials):
+    """ Returns the products details for the getsmarter API """
+    method = 'GET'
+    products = []
+    getsmarter_api_client = GetSmarterEnterpriseApiClient(
+        client_id=getsmarter_client_credentials['CLIENT_ID'],
+        client_secret=getsmarter_client_credentials['CLIENT_SECRET'],
+        provider_url=getsmarter_client_credentials['PROVIDER_URL'], api_url=getsmarter_client_credentials['API_URL']
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/utils.py]: Why are we still sending these settings through arguments? Aren't these settings available from edx-internal?

File: course_discovery/apps/course_metadata/tests/test_utils.py [hunk 9]
```diff
@@ -896,3 +899,133 @@ def test_download_and_save_course_image__for_organization_logo_override_using_dr
...
+                    "stat1Blurb": "<p>A vast number of special beings take this course</p>",
+                    "stat2": "100 million",
+                    "stat2Blurb": "<p>VC fund</p>"
+                }
+            },
+        ]
+    }
+
+    def setUp(self):
+        self.getsmarter_client_credentials = settings.GETSMARTER_CLIENT_CREDENTIALS
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/tests/test_utils.py]: Suggestion: We can make separate dict in `settings/test.py`

File: course_discovery/apps/course_metadata/utils.py (hunk 8 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/utils.py]: You can use them directly from `settings.GETSMARTER_CLIENT_CREDENTIALS.CLIENT_ID`

**Review by Ali-D-Akbar — CHANGES_REQUESTED** [review]
Some changes that can be done to improve the code quality.

File: course_discovery/apps/course_metadata/tests/test_utils.py (hunk 9 above)
Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/tests/test_utils.py]: Done

File: course_discovery/apps/course_metadata/utils.py (hunk 8 above)
Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py]: Removed client credentials from functional and passed them directly to the functiom
`settings..GETSMARTER_CLIENT_CREDENTIALS` is a dict so i think we can't refer its keys like object attribute

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/utils.py]: Ah yes, it's a dict so you can use it in a `dict['key']` way.

File: course_discovery/apps/course_metadata/tests/test_utils.py (hunk 9 above)
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tests/test_utils.py]: nit: We dont need self.__ variable. Use settings variable directly where needed.

File: course_discovery/apps/course_metadata/utils.py [hunk 10]
```diff
@@ -882,3 +883,31 @@ def transform_skills_data(skills_data):
         }
         skills.append(skill_dict)
     return skills
+
+
+def fetch_getsmarter_products():
+    """ Returns the products details from the getsmarter API """
+    method = 'GET'
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py]: no need to define this special variable.

File: course_discovery/settings/test.py [hunk 11]
```diff
@@ -86,3 +86,11 @@
...
 
 DEGREE_CONTENTFUL_CONTENT_TYPE = 'degreeDetailPage'
+
+GETSMARTER_CLIENT_CREDENTIALS = {
+    'CLIENT_ID' : 'test_id',
+    'CLIENT_SECRET' : 'test_secret',
+    'API_URL' : 'https://test-getsmarter.com/api/v1',
+    'PROVIDER_URL' : 'https://auth-test.com',
+    'PRODUCTS_DETAILS_URL' : 'https://test-getsmarter.com/api/v1/products?detail=2',
+}
```
Comment (DawoudSheraz) [inline course_discovery/settings/test.py]: missing end of file line.

**Review by DawoudSheraz — APPROVED** [review]
- A few nit comments
- Please update PR description to include the tests of skipped test

**Review by Ali-D-Akbar — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@AfaqShuaib09 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/3777&entry.1484655318=https://github.com/AfaqShuaib09&entry.752974735=2023-01-26+20:00&entry.1917517419=2023-02-13+17:12&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**